lib/src/client.dart

```diff
@@ -95,34 +168,166 @@ class ClientConnection {
+        _state != ConnectionState.transientFailure) {
+      return;
+    }
+    _setState(ConnectionState.connecting);
+    connectTransport().then((transport) {
+      _currentReconnectDelay = null;
+      _transport = transport;
+      _transport.onActiveStateChanged = _handleActiveStateChanged;
+      _setState(ConnectionState.ready);
+      _pendingCalls.forEach(_startCall);
+      _pendingCalls.clear();
+    }).catchError(_handleTransientFailure);
+  }
+
+  void dispatchCall(ClientCall call) {
+    switch (_state) {
+      case ConnectionState.ready:
+        _startCall(call);
+        break;
+      case ConnectionState.shutdown:
+        _shutdownCall(call);
+        break;
+      default:
+        _pendingCalls.add(call);
+        if (_state == ConnectionState.idle) {
+          _connect();
+        }
+    }
+  }
+
+  void _startCall(ClientCall call) {
+    if (call._isCancelled) return;
+    final headers =
+        createCallHeaders(options.isSecure, authority, call.path, call.options);
+    final stream = _transport.makeRequest(headers);
+    call._onConnectedStream(stream);
+  }
+
+  void _shutdownCall(ClientCall call) {
+    if (call._isCancelled) return;
+    call._onConnectError(
+        new GrpcError.unavailable('Connection shutting down.'));
+  }
+
   /// Shuts down this connection.
   ///
   /// No further calls may be made on this connection, but existing calls
   /// are allowed to finish.
   Future<Null> shutdown() {
-    // TODO(jakobr): Manage streams, close [_transport] when all are done.
-    return _transport.finish();
+    if (_state == ConnectionState.shutdown) return new Future.value();
+    _setShutdownState();
+    return _transport?.finish() ?? new Future.value();
   }
 
   /// Terminates this connection.
   ///
   /// All open calls are terminated immediately, and no further calls may be
   /// made on this connection.
   Future<Null> terminate() {
-    // TODO(jakobr): Manage streams, close them immediately.
-    return _transport.terminate();
+    _setShutdownState();
+    return _transport?.terminate() ?? new Future.value();
   }
 
-  /// Starts a new RPC on this connection.
-  ///
-  /// Creates a new transport stream on this connection, and sends initial call
-  /// metadata.
-  ClientTransportStream sendRequest(
-      bool useTls, String authority, String path, CallOptions options) {
-    final headers = createCallHeaders(useTls, authority, path, options);
-    final stream = _transport.makeRequest(headers);
-    // TODO(jakobr): Manage streams. Subscribe to stream state changes.
-    return stream;
+  void _setShutdownState() {
+    _setState(ConnectionState.shutdown);
+    _cancelTimer();
+    _pendingCalls.forEach(_shutdownCall);
+    _pendingCalls.clear();
+  }
+
+  void _setState(ConnectionState state) {
+    _state = state;
+    if (onStateChanged != null) {
+      onStateChanged(this);
+    }
+  }
+
+  void _handleIdleTimeout() {
+    if (_timer == null) return;
+    _timer = null;
```

Comment:
Use `_cancelTimer` instead of setting `_timer` to `null` in more than one place. Conveys intent better, I think. Plus it cancels unwanted timer events instead of attempting to ignore them when they occur. Unwanted timer events can make debugging hard, but I think your attempt at ignoring them may even fail, if a new timer has been set up in the mean time? You may need to receive the timer object with timer events and compare to current `_timer` before reacting.

Reply:
I've added in a check for the current state. Idle timeout should only fire while in the `ready` state. We only set the timeout when there are no more active streams, and we explicitly clear it when the connection is active again, so if we get a timeout callback, and:

* _state == ready
* _timer != null

then at least there are no active streams. Worst case, we end up closing an idle connection too soon, but the following would have to happen:

1. Connection has been idle long enough for the timer to fire and schedule an event to call `_handleTimeout()`.
1. Before `_handleTimeout()` is actually called, the currently running code makes an RPC, making the connection active (and clearing `_timer` in the process).
1. Still before the scheduled `_handleTimeout()` call is invoked, the RPC manages to finish, and a new idle timeout is scheduled.
1. The original `_handleTimeout()` call finally runs, and closes the connection.

I don't think it's possible to complete an RPC without processing other events in the queue (especially since the `onActiveStateChange` call to mark the transport idle is delivered in a Future event, which would come later in the queue), so I we should be good now.